theurgy.lic

```diff
@@ -2,185 +2,453 @@
 =end
 
-custom_require.call(%w[common common-items common-money common-travel drinfomon spellmonitor equipmanager])
+custom_require.call(%w[common common-arcana common-items common-money common-travel drinfomon spellmonitor equipmanager])
 
-class Theurgy
+class TheurgyActions
   include DRC
   include DRCI
-  include DRCM
   include DRCT
+  include DRCM
+  include DRCA
+  
+  def initialize(settings, data)
+    @settings = settings
+    @hometown = @settings.hometown
+    @data = data
+
+    @theurgy_supply_container = @settings.theurgy_supply_container
+    @water_holder = @settings.water_holder
+    @immortal_aspect = @settings.immortal_aspect
+    @flint_lighter = @settings.flint_lighter
+    @safe_room = @settings.safe_room
+
+    @rituals = get_rituals
+    @communes = get_communes
+
+    annotate_item_presence(@rituals + @communes)
+    
+    @rituals.reject! do |r|
+      r[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+    @communes.reject! do |c|
+      c[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+  end
 
-  ######################## GET/PUT ########################
-
-  def put_in_container(item)
-    bput "put #{item} in my #{@theurgy_supply_container}", 'You put', 'What were you referring to'
+  def get_rituals
+    rituals = all_rituals
+    blacklist = @settings.theurgy_blacklist
+    rituals.reject! { |r| blacklist.include? r[:method].to_s }
+    whitelist = @settings.theurgy_whitelist
+    unless whitelist.empty?
+      rituals.select! { |r| whitelist.include? r[:method].to_s }
+    end
+    rituals
+  end
+  
+  def all_rituals
+    rituals = []
+    if @data['almsbox'] && @settings.tithe
+      rituals << {
+        method: :tithe,
+      }
+    end
+    if @hometown == 'Crossing'
+      if DRStats.circle >= 30
+        rituals << {
+          method: :study_wall,
+        }
+      end
+      rituals << {
+        method: :refectory,
+      }
+    end
+    if DRSkill.getrank('Mechanical Lore') > 140 && @immortal_aspect && @data['limb_foraging']
+      # TODO: allow other knife nouns 
+      rituals << {
+        method: :carve_bead,
+        items: [
+          { name: 'carving knife', any_container: true },
+          { name: 'shaper', any_container: true},
+          { name: 'prayer chain', any_container: true },
+        ],
+      }
+    end
+    if @data['gather_sirese'] && @data['plant_sirese']
+      rituals << {
+        method: :sirese_seed,
+      }
+    end
+    if @data['bath']
+      herbs = @data['herbs'] || ['sage', 'lavender']
+      rituals << {
+        method: :bathe,
+        items: herbs.map { |h| { name: h, shop: @data['theurgy_supplies'] } },
+      }
+    end
+    if @immortal_aspect
+      ritals << {
+        method: :meditate_bead,
+        items: [
+          { name: 'prayer chain', any_container: true },
+        ],
+      }
+    end
+    rituals << {
+      method: :pray_badge,
+    }
+    # TODO: prayer mat support
+    if @data['altar']
```

Comment:
Some altars provide various things that can be done at them. I know there's a list online that let's you see what can be done at what altar. Was this checked against that? These all look standard, just don't know for sure.

Reply:
I've tested these actions at each of these altars.

Reply:
OK, thanks!